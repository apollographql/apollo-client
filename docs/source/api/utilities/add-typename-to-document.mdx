---
title: addTypenameToDocument
description: Adds __typename fields to all selection sets in a GraphQL document
---

<DocBlock
  canonicalReference="@apollo/client!~addTypenameToDocument:var"
  customOrder={["summary", "remarks", "example"]}
/>

## Function signature

```ts
function addTypenameToDocument<TNode extends ASTNode>(doc: TNode): TNode;
```

