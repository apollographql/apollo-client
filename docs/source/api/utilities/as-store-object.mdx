---
title: AsStoreObject
description: Type that adds an implicit index signature to make types assignable to StoreObject
---

<InterfaceDetails
  canonicalReference="@apollo/client/utilities!AsStoreObject:type"
  headingLevel={2}
  displayName="AsStoreObject"
/>