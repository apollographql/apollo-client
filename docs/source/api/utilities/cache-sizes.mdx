---
title: cacheSizes
description: Global cache size configuration for Apollo Client internal caches
---

<DocBlock
  canonicalReference="@apollo/client/utilities!cacheSizes:variable"
  customOrder={["summary", "remarks", "example"]}
/>

## Types

<InterfaceDetails
  canonicalReference="@apollo/client/utilities!CacheSizes:interface"
  headingLevel={3}
  displayName="CacheSizes"
/>