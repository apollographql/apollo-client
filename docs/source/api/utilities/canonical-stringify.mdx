---
title: canonicalStringify
description: Serializes a value to JSON with object keys in a consistent, sorted order
---

<DocBlock
  canonicalReference="@apollo/client/utilities!canonicalStringify:function"
  customOrder={["summary", "remarks", "example"]}
/>

## Function signature

```ts
function canonicalStringify(value: any): string
```