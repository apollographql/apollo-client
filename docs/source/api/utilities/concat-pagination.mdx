---
title: concatPagination
description: A field policy that concatenates new results onto existing arrays
---

<DocBlock
  canonicalReference="@apollo/client/utilities!concatPagination:function"
  customOrder={["summary", "remarks", "example"]}
/>

## Function signature

```ts
function concatPagination<T = Reference>(
  keyArgs?: KeyArgs
): FieldPolicy<T[]>
```