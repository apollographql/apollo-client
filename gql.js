// This is here for backcompat, even though the right way is to use the other package
module.exports = require('graphql-tag');
// We emit a development-only warning to deprecate this import
if(process.env.NODE_ENV === 'development') {
  console.warn(
    "Requiring 'apollo-client/gql' is now deprecated. " +
    "You should require 'graphql-tag' instead. " +
    "See http://docs.apollostack.com/apollo-client/core.html#gql for more information."
  )
}
