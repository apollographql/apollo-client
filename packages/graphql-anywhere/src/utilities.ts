import { DocumentNode, DirectiveNode } from 'graphql';

import { getInclusionDirectives } from 'apollo-utilities';

import { graphql, VariableMap, ExecInfo, ExecContext } from './graphql';

import { invariant } from 'ts-invariant';

const { hasOwnProperty } = Object.prototype;

export function filter<FD = any, D extends FD = any>(
  doc: DocumentNode,
  data: D,
  variableValues: VariableMap = {},
): FD {
  if (data === null) return data;

  const resolver = (
    fieldName: string,
    root: any,
    args: Object,
    context: ExecContext,
    info: ExecInfo,
  ) => {
    return root[info.resultKey];
  };

  return Array.isArray(data)
    ? data.map(dataObj => graphql(resolver, doc, dataObj, null, variableValues))
    : graphql(resolver, doc, data, null, variableValues);
}

// TODO: we should probably make check call propType and then throw,
// rather than the other way round, to avoid constructing stack traces
// for things like oneOf uses in React. At this stage I doubt many people
// are using this like that, but in the future, who knows?
export function check(
  doc: DocumentNode,
  data: any,
  variables: VariableMap = {},
): void {
  const resolver = (
    fieldName: string,
    root: any,
    args: any,
    context: any,
    info: any,
  ) => {
    // When variables is null, fields with @include/skip directives that
    // reference variables are considered optional.
    invariant(
      hasOwnProperty.call(root, info.resultKey) ||
        (!variables && hasVariableInclusions(info.field.directives)),
      `${info.resultKey} missing on ${JSON.stringify(root)}`,
    );
    return root[info.resultKey];
  };

  graphql(resolver, doc, data, {}, variables, {
    fragmentMatcher: () => false,
  });
}

function hasVariableInclusions(
  directives: ReadonlyArray<DirectiveNode>,
): boolean {
  return getInclusionDirectives(directives).some(
    ({ ifArgument }) =>
      ifArgument.value && ifArgument.value.kind === 'Variable',
  );
}

// Lifted/adapted from
//   https://github.com/facebook/react/blob/master/src/isomorphic/classic/types/ReactPropTypes.js
const ANONYMOUS = '<<anonymous>>';
function PropTypeError(message) {
  this.message = message;
  this.stack = '';
}
// Make `instanceof Error` still work for returned errors.
PropTypeError.prototype = Error.prototype;

const reactPropTypeLocationNames = {
  prop: 'prop',
  context: 'context',
  childContext: 'child context',
};

function createChainableTypeChecker(validate) {
  function checkType(
    isRequired,
    props,
    propName,
    componentName,
    location,
    propFullName,
  ) {
    componentName = componentName || ANONYMOUS;
    propFullName = propFullName || propName;
    if (props[propName] == null) {
      const locationName = reactPropTypeLocationNames[location];
      if (isRequired) {
        if (props[propName] === null) {
          return new PropTypeError(
            `The ${locationName} \`${propFullName}\` is marked as required ` +
              `in \`${componentName}\`, but its value is \`null\`.`,
          );
        }
        return new PropTypeError(
          `The ${locationName} \`${propFullName}\` is marked as required in ` +
            `\`${componentName}\`, but its value is \`undefined\`.`,
        );
      }
      return null;
    } else {
      return validate(props, propName, componentName, location, propFullName);
    }
  }

  const chainedCheckType = checkType.bind(null, false);
  chainedCheckType.isRequired = checkType.bind(null, true);

  return chainedCheckType;
}

export function propType(
  doc: DocumentNode,
  mapPropsToVariables = props => null,
) {
  return createChainableTypeChecker((props, propName) => {
    const prop = props[propName];
    try {
      if (!prop.loading) {
        check(doc, prop, mapPropsToVariables(props));
      }
      return null;
    } catch (e) {
      // Need a much better error.
      // Also we aren't checking for extra fields
      return e;
    }
  });
}
