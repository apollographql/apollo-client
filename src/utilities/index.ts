export {
  DirectiveInfo,
  InclusionDirectives,
  shouldInclude,
  hasDirectives,
  hasClientExports,
  getDirectiveNames,
  getInclusionDirectives,
} from './graphql/directives';

export {
  FragmentMap,
  createFragmentMap,
  getFragmentQueryDocument,
  getFragmentFromSelection,
} from './graphql/fragments';

export {
  checkDocument,
  getOperationDefinition,
  getOperationName,
  getFragmentDefinitions,
  getQueryDefinition,
  getFragmentDefinition,
  getMainDefinition,
  getDefaultValues,
} from './graphql/getFromAST';

export {
  Reference,
  StoreValue,
  Directives,
  VariableValue,
  makeReference,
  isReference,
  isField,
  isInlineFragment,
  valueToObjectRepresentation,
  storeKeyNameFromField,
  argumentsObjectFromField,
  resultKeyNameFromField,
  getStoreKeyName,
  getTypenameFromResult,
} from './graphql/storeUtils';

export {
  RemoveNodeConfig,
  GetNodeConfig,
  RemoveDirectiveConfig,
  GetDirectiveConfig,
  RemoveArgumentsConfig,
  GetFragmentSpreadConfig,
  RemoveFragmentSpreadConfig,
  RemoveFragmentDefinitionConfig,
  RemoveVariableDefinitionConfig,
  addTypenameToDocument,
  buildQueryFromSelectionSet,
  removeDirectivesFromDocument,
  removeConnectionDirectiveFromDocument,
  removeArgumentsFromDocument,
  removeFragmentSpreadFromDocument,
  removeClientSetsFromDocument,
} from './graphql/transform';
