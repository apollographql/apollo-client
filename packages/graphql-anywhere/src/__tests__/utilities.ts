import gql, { disableFragmentWarnings } from 'graphql-tag';
import { checkPropTypes } from 'prop-types';

// Turn off warnings for repeated fragment names
disableFragmentWarnings();

import { filter, check, propType } from '../utilities';

describe('utilities', () => {
  describe('with a single query', () => {
    const doc = gql`
      {
        alias: name
        height(unit: METERS)
        avatar {
          square
        }
      }
    `;
    const fragment = gql`
      fragment foo on Foo {
        alias: name
        height(unit: METERS)
        avatar {
          square
        }
      }
    `;
    const fragmentWithAVariable = gql`
      fragment foo on Foo {
        alias: name
        height(unit: METERS)
        avatar @include(if: $foo) {
          square
        }
      }
    `;
    const data = {
      alias: 'Bob',
      name: 'Wrong',
      height: 1.89,
      avatar: {
        square: 'abc',
        circle: 'def',
        triangle: 'qwe',
      },
    };
    const filteredData = {
      alias: 'Bob',
      height: 1.89,
      avatar: {
        square: 'abc',
      },
    };
    const arrayData = [
      {
        alias: 'Bob',
        name: 'Wrong',
        height: 1.89,
        avatar: {
          square: 'abc',
          circle: 'def',
          triangle: 'qwe',
        },
      },
      {
        alias: 'Tom',
        name: 'Right',
        height: 1.77,
        avatar: {
          square: 'jkl',
          circle: 'bnm',
          triangle: 'uio',
        },
      },
    ];
    const filteredArrayData = [
      {
        alias: 'Bob',
        height: 1.89,
        avatar: {
          square: 'abc',
        },
      },
      {
        alias: 'Tom',
        height: 1.77,
        avatar: {
          square: 'jkl',
        },
      },
    ];

    beforeEach(() => {
      checkPropTypes.resetWarningCache();
      jest.spyOn(global.console, 'error').mockImplementation(() => {});
    });

    afterEach(() => {
      global.console.error.mockRestore();
    });

    it('can filter data', () => {
      expect(filter(doc, data)).toEqual(filteredData);
    });

    it('can filter an array of data', () => {
      expect(filter(doc, arrayData)).toEqual(filteredArrayData);
    });

    it('can filter data for fragments ', () => {
      expect(filter(fragment, data)).toEqual(filteredData);
    });

    it('can filter data for fragments with variables', () => {
      expect(filter(fragmentWithAVariable, data, { foo: true })).toEqual(
        filteredData,
      );
    });

    it('can generate propTypes for fragments', () => {
      expect(propType(fragment)).toEqual(expect.any(Function));
    });

    it('can check propTypes for fragments', () => {
      const propTypes = {
        foo: propType(fragment),
      };
      checkPropTypes(propTypes, filteredData, 'prop', 'MyComponent');
      expect(global.console.error).not.toHaveBeenCalled();
      checkPropTypes(propTypes, { foo: {} }, 'prop', 'MyComponent');
      expect(global.console.error.mock.calls[0]).toMatchSnapshot();
    });

    it('can generate propTypes for fragments with variables', () => {
      expect(propType(fragmentWithAVariable)).toEqual(expect.any(Function));
    });

    it('can check propTypes for fragments with variables', () => {
      const mapPropsToVariables = () => null;
      const propTypes = {
        foo: propType(fragmentWithAVariable, mapPropsToVariables),
      };
      checkPropTypes(propTypes, { foo: filteredData }, 'prop', 'MyComponent');
      expect(global.console.error).not.toHaveBeenCalled();
      const badProps = { foo: { ...filteredData } };
      delete badProps.foo.height;
      checkPropTypes(propTypes, badProps, 'prop', 'MyComponent');
      expect(global.console.error).toHaveBeenCalled();
      expect(global.console.error.mock.calls[0]).toMatchSnapshot();
    });

    it('makes variable inclusion props optional, when no variables are passed', () => {
      const propTypes = {
        foo: propType(fragmentWithAVariable),
      };
      const propsWithoutAvatar = { foo: { ...filteredData } };
      delete propsWithoutAvatar.foo.avatar;
      checkPropTypes(propTypes, propsWithoutAvatar, 'prop', 'MyComponent');
      expect(global.console.error).not.toHaveBeenCalled();
    });

    it('can check matching data', () => {
      check(doc, filteredData);
    });

    it('can check matching data for fragments with variables', () => {
      check(doc, filteredData, { foo: true });
    });

    it('throws when checking non-matching data for fragments with variables', () => {
      const badFilteredData = { ...filteredData };
      delete badFilteredData.avatar;
      expect(() => {
        check(doc, badFilteredData);
      }).toThrow();
    });

    // This doesn't throw but potentially it should?
    it('can check overspecified data', () => {
      check(doc, data);
    });

    it('throws when checking underspecified data', () => {
      expect(() => {
        check(doc, {
          name: 'Wrong',
        });
      }).toThrow();

      expect(() => {
        check(doc, {
          alias: 'Bob',
          height: 1.89,
        });
      }).toThrow();
    });
  });

  describe('with a single fragment', () => {
    const doc = gql`
      fragment PersonDetails on Person {
        alias: name
        height(unit: METERS)
        avatar {
          square
        }
      }
    `;
    const data = {
      alias: 'Bob',
      name: 'Wrong',
      height: 1.89,
      avatar: {
        square: 'abc',
        circle: 'def',
        triangle: 'qwe',
      },
    };
    const filteredData = {
      alias: 'Bob',
      height: 1.89,
      avatar: {
        square: 'abc',
      },
    };
    const arrayData = [
      {
        alias: 'Bob',
        name: 'Wrong',
        height: 1.89,
        avatar: {
          square: 'abc',
          circle: 'def',
          triangle: 'qwe',
        },
      },
      {
        alias: 'Tom',
        name: 'Right',
        height: 1.77,
        avatar: {
          square: 'jkl',
          circle: 'bnm',
          triangle: 'uio',
        },
      },
    ];
    const filteredArrayData = [
      {
        alias: 'Bob',
        height: 1.89,
        avatar: {
          square: 'abc',
        },
      },
      {
        alias: 'Tom',
        height: 1.77,
        avatar: {
          square: 'jkl',
        },
      },
    ];

    it('can filter data', () => {
      expect(filter(doc, data)).toEqual(filteredData);
    });

    it('can filter an array of data', () => {
      expect(filter(doc, arrayData)).toEqual(filteredArrayData);
    });

    it('can check matching data', () => {
      check(doc, filteredData);
    });

    // This doesn't throw but potentially it should?
    it('can check overspecified data', () => {
      check(doc, data);
    });

    it('throws when checking underspecified data', () => {
      expect(() => {
        check(doc, {
          name: 'Wrong',
        });
      }).toThrow();

      expect(() => {
        check(doc, {
          alias: 'Bob',
          height: 1.89,
        });
      }).toThrow();
    });
  });

  describe('with a single fragment', () => {
    const doc = gql`
      fragment PersonDetails on Person {
        alias: name
        height(unit: METERS)
        avatar {
          square
        }
      }
    `;
    const data = {
      alias: 'Bob',
      name: 'Wrong',
      height: 1.89,
      avatar: {
        square: 'abc',
        circle: 'def',
        triangle: 'qwe',
      },
    };
    const filteredData = {
      alias: 'Bob',
      height: 1.89,
      avatar: {
        square: 'abc',
      },
    };
    const arrayData = [
      {
        alias: 'Bob',
        name: 'Wrong',
        height: 1.89,
        avatar: {
          square: 'abc',
          circle: 'def',
          triangle: 'qwe',
        },
      },
      {
        alias: 'Tom',
        name: 'Right',
        height: 1.77,
        avatar: {
          square: 'jkl',
          circle: 'bnm',
          triangle: 'uio',
        },
      },
    ];
    const filteredArrayData = [
      {
        alias: 'Bob',
        height: 1.89,
        avatar: {
          square: 'abc',
        },
      },
      {
        alias: 'Tom',
        height: 1.77,
        avatar: {
          square: 'jkl',
        },
      },
    ];

    it('can filter data', () => {
      expect(filter(doc, data)).toEqual(filteredData);
    });

    it('can filter an array of data', () => {
      expect(filter(doc, arrayData)).toEqual(filteredArrayData);
    });

    it('can check matching data', () => {
      check(doc, filteredData);
    });

    // This doesn't throw but potentially it should?
    it('can check overspecified data', () => {
      check(doc, data);
    });

    it('throws when checking underspecified data', () => {
      expect(() => {
        check(doc, {
          name: 'Wrong',
        });
      }).toThrow();

      expect(() => {
        check(doc, {
          alias: 'Bob',
          height: 1.89,
        });
      }).toThrow();
    });
  });

  describe('with nested fragments', () => {
    const doc = gql`
      fragment PersonDetails on Person {
        alias: name
        height(unit: METERS)
        avatar {
          square
          ... on Avatar {
            circle
          }
        }
      }
    `;
    const data = {
      alias: 'Bob',
      name: 'Wrong',
      height: 1.89,
      avatar: {
        square: 'abc',
        circle: 'def',
        triangle: 'qwe',
      },
    };
    const filteredData = {
      alias: 'Bob',
      height: 1.89,
      avatar: {
        square: 'abc',
        circle: 'def',
      },
    };
    const arrayData = [
      {
        alias: 'Bob',
        name: 'Wrong',
        height: 1.89,
        avatar: {
          square: 'abc',
          circle: 'def',
          triangle: 'qwe',
        },
      },
      {
        alias: 'Tom',
        name: 'Right',
        height: 1.77,
        avatar: {
          square: 'jkl',
          circle: 'bnm',
          triangle: 'uio',
        },
      },
    ];
    const filteredArrayData = [
      {
        alias: 'Bob',
        height: 1.89,
        avatar: {
          square: 'abc',
          circle: 'def',
        },
      },
      {
        alias: 'Tom',
        height: 1.77,
        avatar: {
          square: 'jkl',
          circle: 'bnm',
        },
      },
    ];

    it('can filter data', () => {
      expect(filter(doc, data)).toEqual(filteredData);
    });

    it('can filter an array of data', () => {
      expect(filter(doc, arrayData)).toEqual(filteredArrayData);
    });

    it('can check matching data', () => {
      check(doc, filteredData);
    });

    // This doesn't throw but potentially it should?
    it('can check overspecified data', () => {
      check(doc, data);
    });

    it('throws when checking underspecified data', () => {
      expect(() => {
        check(doc, {
          name: 'Wrong',
        });
      }).toThrow();

      expect(() => {
        check(doc, [
          {
            name: 'Wrong',
          },
        ]);
      }).toThrow();

      expect(() => {
        check(doc, {
          alias: 'Bob',
          height: 1.89,
        });
      }).toThrow();

      expect(() => {
        check(doc, {
          alias: 'Bob',
          height: 1.89,
          avatar: {
            // missing the correct field
            triangle: 'qwe',
          },
        });
      }).toThrow();
    });

    describe('if the nested fragment has not matched', () => {
      it('can filter data', () => {
        const filtered = filter(doc, {
          alias: 'Bob',
          name: 'Wrong',
          height: 1.89,
          avatar: {
            square: 'abc',
            // there is no circle field here, but we can't know if that's not
            // because avatar is not an Avatar
            triangle: 'qwe',
          },
        });

        expect(filtered).toEqual({
          alias: 'Bob',
          height: 1.89,
          avatar: {
            square: 'abc',
          },
        });
      });

      it('does not throw when checking', () => {
        check(doc, {
          alias: 'Wrong',
          height: 1.89,
          avatar: {
            square: 'abc',
            // there is no circle field here, but we can't know if that's not
            // because avatar is not an Avatar
          },
        });
      });
    });
  });
});
