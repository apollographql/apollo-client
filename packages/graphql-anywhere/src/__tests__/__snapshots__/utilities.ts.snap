// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilities with a single query can check propTypes for fragments 1`] = `
Array [
  "Warning: Failed prop type: alias missing on {}",
]
`;

exports[`utilities with a single query can check propTypes for fragments with variables 1`] = `
Array [
  "Warning: Failed prop type: height missing on {\\"alias\\":\\"Bob\\",\\"avatar\\":{\\"square\\":\\"abc\\"}}",
]
`;
